cpp/src/arrow/compute/kernels/argsort.cc

```diff
@@ -0,0 +1,189 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/argsort.h"
+#include <algorithm>
+#include <numeric>
+#include <vector>
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/expression.h"
+#include "arrow/compute/logical_type.h"
+#include "arrow/type_traits.h"
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+/// \brief UnaryKernel implementing Argsort operation
+class ARROW_EXPORT ArgsortKernel : public UnaryKernel {
+ protected:
+  std::shared_ptr<DataType> type_;
+
+ public:
+  /// \brief UnaryKernel interface
+  ///
+  /// delegates to subclasses via Argsort()
+  Status Call(FunctionContext* ctx, const Datum& values, Datum* offsets) override = 0;
+
+  /// \brief output type of this kernel
+  std::shared_ptr<DataType> out_type() const override { return uint64(); }
+
+  /// \brief single-array implementation
+  virtual Status Argsort(FunctionContext* ctx, const std::shared_ptr<Array>& values,
+                         std::shared_ptr<Array>* offsets) = 0;
+
+  /// \brief factory for ArgsortKernel
+  ///
+  /// \param[in] value_type constructed ArgsortKernel will support sorting
+  ///            values of this type
+  /// \param[out] out created kernel
+  static Status Make(const std::shared_ptr<DataType>& value_type,
+                     std::unique_ptr<ArgsortKernel>* out);
+};
+
+template <typename ArrayType>
+bool CompareValues(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.Value(lhs) < array.Value(rhs);
+}
+
+template <typename ArrayType>
+bool CompareViews(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.GetView(lhs) < array.GetView(rhs);
+}
+
+template <typename ArrowType, typename Comparator>
+class ArgsortKernelImpl : public ArgsortKernel {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+ private:
+  Comparator compare;
```

Comment:
Code style:
```suggestion
  Comparator compare_;
```

Additionally, please reorder this class to be a public section followed by a private section